File: pump/pump_system.cpp
// Notes: Entire system runs at 1.3A. The flow rate of the system is about 165mL/min.
// To have 2.25L of water processed, about 13.6 min is required. To have all the filters
// used, a total of 40.9 min is needed. The system requires about 0.886 Ah of total charge.
// At least 1Ah is recommended.
#include <iostream>
// for delay function.
#include <chrono>
#include <string>
#include <thread>
#include <cstdio>

// for signal handling
#include <JetsonGPIO.h>
#include <signal.h>

// namespace

using namespace std;

// BOARD Pin Definitions
const int pump_pin = 12;
const int but_pin = 18;   
const int flow_pin = 16;
const int filter_1_pin = 11;
const int filter_2_pin = 13;
const int filter_3_pin = 15;

// GLOBAL VARS

static int user_input_filter = 0;
static bool end_this_program = false;
static volatile int pulse_count = 0;
static volatile bool pump = 0;

// USEFULL FUNCTIONS

inline void delayMs(int ms) { this_thread::sleep_for(chrono::milliseconds(ms)); }

void signalHandler(int s) { end_this_program = true; }

// INTERRUPT FUNCTIONS

// Toggles the pump. Outputs the measured amount of water when between on/off toggles.
// Count is reset when
void pumpToggle(const std::string& channel)
{
    
    pump = !pump; // toggle state
    if (pump) {
        puts("PUMP ON");
        pulse_count = 0;
        GPIO::output(user_input_filter, GPIO::HIGH);
        GPIO::output(pump_pin, GPIO::HIGH);
    } else {
        puts("PUMP OFF!");
        GPIO::output(user_input_filter, GPIO::LOW);
        GPIO::output(pump_pin, GPIO::LOW);
        cout << pulse_count*0.223 << " mL" << endl;
    }
        
}

// ~4.467 pulses/mL -> 0.223 mL/pulse
void flowHandler(const std::string& channel)
{
    pulse_count++;
}

// get an input for activating the filter to collect from
int getInputFilter()
{
    static string input;
    do
    {
        puts("Enter filter number to activate! (1, 2, 3)");
        cin >> input;
        if (input == "1")
                return filter_1_pin;
        else if (input == "2")
                return filter_2_pin;
        else if (input == "3")
                return filter_3_pin;

    } while (!end_this_program);

}

int main()
{
    // When CTRL+C pressed, signalHandler will be called
    signal(SIGINT, signalHandler);

    // Pin Setup.
    GPIO::setmode(GPIO::BOARD);

    // set pin as an output pin with optional initial state of HIGH
    GPIO::setup(pump_pin, GPIO::OUT, GPIO::LOW);
    GPIO::setup(but_pin, GPIO::IN);
    GPIO::setup(flow_pin, GPIO::IN);
    GPIO::setup(filter_1_pin, GPIO::OUT, GPIO::LOW);
    GPIO::setup(filter_2_pin, GPIO::OUT, GPIO::LOW);
    GPIO::setup(filter_3_pin, GPIO::OUT, GPIO::LOW);

    cout << "Starting demo now! Press CTRL+C to exit" << endl;

    // detect events and trigger function
    // GPIO::add_event_detect(but_pin, GPIO::Edge::RISING, pumpToggle, 20);
    GPIO::add_event_detect(flow_pin, GPIO::Edge::RISING, flowHandler, 0);
    try {
        while (!end_this_program)
        {

            user_input_filter = getInputFilter(); // get input from user until it's valid or program ends
            GPIO::add_event_detect(but_pin, GPIO::Edge::RISING, pumpToggle, 200);
            puts("Push button to start!");
            while (pump == 0) {};
            puts("Push button to end!");
            while (pump == 1) {};
            GPIO::remove_event_detect(but_pin);
        }
    } catch (exception& exc) {
        cout << exc.what() << endl;
    }
    GPIO::cleanup();

    return 0;
}

File: pump/pump_system.py
import RPi.GPIO as GPIO
import time

class pump_system:
    def __init__(self):
        # VARIABLES (DO NOT CHANGE UNLESS NECESSARY)
        self.input_filter = 1                    # Tracks which filter to activate
        self.pulse_count = 0                     # Used to calculate how much water was collected
        self.water_sampled = 0                   # Stores how much water was sampled in the last toggle
        self.pump = False                        # Stores state of the pump
        self.mLPerPulse = 0.223                  # Stores mL per pulse, can be changed here
        
        # PIN DEFINITIONS (Can be redefined here)
        self.pump_pin = 12
        self.flow_pin = 16
        self.filter_1_pin = 11
        self.filter_2_pin = 13
        self.filter_3_pin = 15
        
        self.setUp() # Call function to setup pins

    # INTERRUPT HANDLERS

    # toggles pump and specific valve solenoid to select filter
    def toggleCollection(self, channel=None):
        
        self.pump = ~self.pump
        if (self.pump):
            # PUMP ON
            self.pulse_count = 0
            GPIO.output(self.pump_pin, GPIO.HIGH)
            GPIO.output(self.input_filter, GPIO.HIGH)
        else:
            # PUMP OFF
            GPIO.output(self.pump_pin, GPIO.LOW)
            time.sleep(2)
            GPIO.output(self.input_filter, GPIO.LOW)
            self.water_sampled = self.pulse_count*self.mLPerPulse

    # Counts number of pulses
    def flowHandler(self, channel):
        self.pulse_count += 1

    # METHODS

    # Sets up all of the GPIO pins required for the pump system
    def setUp(self):
        GPIO.setmode(GPIO.BOARD)
        GPIO.setup(self.flow_pin, GPIO.IN)
        GPIO.setup(self.pump_pin, GPIO.OUT)
        GPIO.setup(self.filter_1_pin, GPIO.OUT)
        GPIO.setup(self.filter_2_pin, GPIO.OUT)
        GPIO.setup(self.filter_3_pin, GPIO.OUT)
        GPIO.add_event_detect(self.flow_pin, GPIO.RISING, callback=self.flowHandler, bouncetime=0)

# Triggers collection of 2L sample
    def collectSample(self, filter):
        print("Starting Collection of 2L!")
        total_sampled = 0 # in mL
        duration = 0      # in seconds
        self.setInputFilter(filter)
        while (total_sampled < 2100):
            self.toggleCollection()
            time.sleep(8)
            self.toggleCollection()
            total_sampled += self.water_sampled
            print("Total Sampled: " + str(total_sampled) + "mL")
            duration += 1

    # Use to properly select filter, otherwise sets to one
    def setInputFilter(self, input_u):
        if (input_u == 1):
            self.input_filter = self.filter_1_pin
        elif (input_u == 2):
            self.input_filter = self.filter_2_pin
        elif(input_u == 3):
            self.input_filter = self.filter_3_pin
        else:
            self.input_filter = self.filter_1_pin

    # Call to cleanup, mainly for testing
    def clean(err=None):
        GPIO.output(self.pump_pin, GPIO.LOW)
        GPIO.output(self.input_filter, GPIO.LOW)
        GPIO.cleanup()

    # DEMO METHODS

    # Used for demo
    def getInputFilter(self):
        while(1):
            input_u = input("Enter filter number to activate! (1, 2, 3)\n")
            if (input_u == 1):
                self.input_filter = self.filter_1_pin
                return
            elif (input_u == 2):
                self.input_filter = self.filter_2_pin
                return
            elif(input_u == 3):
                self.input_filter = self.filter_3_pin
                return

    # Used for demo, requires button, no way out except ctrl+c
    def demo(self):
        print("Starting demo now! Press CTRL+C to exit/stop")
        try:
            while (1):
                self.getInputFilter() # get input from user until it's valid
                self.toggleCollection()
                print("Starting collection! Press CTRL+C to stop")
                try:
                    time.sleep(10)
                except KeyboardInterrupt:
                    pass
                self.toggleCollection()
                print("Water Sampled: " + str(self.water_sampled) + " mL")
        except KeyboardInterrupt:
            self.clean()
            exit(0)








